crates/subspace-networking/src/behavior/persistent_parameters.rs

```diff
@@ -0,0 +1,248 @@
+
+        Ok(Self {
+            db: Arc::new(db),
+            column_id,
+            object_id,
+            known_peers: cache,
+            networking_parameters_save_delay: Self::default_delay(),
+        })
+    }
+
+    // Helps create a copy of the internal LruCache
+    fn clone_known_peers(&self) -> LruCache<PeerId, HashSet<Multiaddr>> {
+        let mut known_peers = LruCache::new(self.known_peers.cap());
+
+        for (peer_id, addresses) in self.known_peers.iter() {
+            known_peers.push(*peer_id, addresses.clone());
+        }
+
+        known_peers
+    }
+
+    /// Creates a reference to the `NetworkingParametersRegistry` trait implementation.
+    pub fn boxed(self) -> Box<dyn NetworkingParametersRegistry> {
+        Box::new(self)
+    }
+
+    // Create default delay for networking parameters.
+    fn default_delay() -> Pin<Box<Fuse<Sleep>>> {
+        Box::pin(sleep(Duration::from_secs(DATA_FLUSH_DURATION_SECS)).fuse())
+    }
+}
+
+#[async_trait]
+impl NetworkingParametersRegistry for NetworkingParametersManager {
+    async fn add_known_peer(&mut self, peer_id: PeerId, addresses: Vec<Multiaddr>) {
+        let addr_set = addresses.iter().cloned().collect::<HashSet<_>>();
+
+        if let Some(addresses) = self.known_peers.get_mut(&peer_id) {
+            *addresses = addresses.union(&addr_set).cloned().collect()
+        } else {
+            self.known_peers.push(peer_id, addr_set);
+        }
+    }
+
+    async fn known_addresses(&self, peer_number: usize) -> Vec<(PeerId, Multiaddr)> {
+        self.known_peers
+            .iter()
+            .take(peer_number)
+            .flat_map(|(peer_id, addresses)| addresses.iter().map(|addr| (*peer_id, addr.clone())))
+            .map(|(peer_id, addr)| {
+                // remove p2p-protocol suffix if any
+                let mut modified_address = addr.clone();
+
+                if let Some(Protocol::P2p(_)) = modified_address.pop() {
+                    (peer_id, modified_address)
+                } else {
+                    (peer_id, addr)
+                }
+            })
+            .collect()
+    }
+
+    async fn run(&mut self) {
+        (&mut self.networking_parameters_save_delay).await;
+
+        // save accumulated cache to DB
+        let dto: NetworkingParametersDto = self.clone_known_peers().into();
+        let save_result = serde_json::to_vec(&dto)
+            .map_err(NetworkParametersPersistenceError::from)
+            .and_then(|data| {
+                let tx = vec![(self.column_id, self.object_id, Some(data))];
+
+                self.db.commit(tx).map_err(|err| err.into())
+            });
+
+        if let Err(err) = save_result {
+            debug!(error=%err, "Error on saving network parameters");
+        } else {
+            trace!("Networking parameters saved to DB");
+        }
+
+        self.networking_parameters_save_delay = NetworkingParametersManager::default_delay();
```

Comment:
This should have been a loop, no? Also this causes DB writes even if nothing has changed.

I'm thinking it would be nice to have a receiver channel here so that rare updates happen instantly, but rare updates are debounced. There is probably a crate for that though.

Reply:
The previous version contained a loop when this was run in a separate worker. Do we need such a logic here? It doesn't seem like a performance critical point, we could increase the delay instead.  Also, `cache_updated` flag seems like an easy fix. Will do.

Reply:
I mean if this is supposed to run forever (until dropped that is), then it should be a loop. It is unexpected for me that it exits right away. In the future we'll likely want to turn all of these `fn run` structs to inherit `Future` trait, then it'll be more obvious why this should have had a loop.